pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java

```diff
@@ -463,4 +490,71 @@ private long getWatermarkMs(long minStartTimeMs, long bucketMs, String mergeLeve
 
     return pinotTaskConfigs;
   }
+
+  private long getMergeRollupTaskDelayInNumTimeBuckets(long watermarkMs, long bufferTimeMs, long bucketTimeMs) {
+    if (bufferTimeMs == 0 || watermarkMs == -1) {
+      return 0;
+    }
+    return (long) Math.floor((System.currentTimeMillis() - watermarkMs - bufferTimeMs) / (double) bucketTimeMs);
+  }
+
+  private void setWatermarkMs(String tableNameWithType, String mergeLevel, long watermarkMs, long bufferTimeMs,
+      long bucketTimeMs) {
+    LOGGER.info(
+        "Setting watermark for table: {} and mergeLevel: {} is {} (watermarkMs={}, bufferTimeMs={}, bucketTimeMs={})",
+        tableNameWithType, mergeLevel, getMergeRollupTaskDelayInNumTimeBuckets(watermarkMs, bufferTimeMs, bucketTimeMs),
+        watermarkMs, bucketTimeMs, bucketTimeMs);
+
+    ControllerMetrics controllerMetrics = _clusterInfoAccessor.getControllerMetrics();
+    if (controllerMetrics == null) {
+      return;
+    }
+
+    PinotMetricsRegistry metricsRegistry = controllerMetrics.getMetricsRegistry();
+    if (metricsRegistry == null) {
+      return;
+    }
+
+    // Update gauge value that indicates the delay in terms of the number of time buckets.
+    _mergeRollupWatermarks.computeIfAbsent(tableNameWithType, k -> new HashMap<>());
+    Map<String, Long> watermarkForTable = _mergeRollupWatermarks.get(tableNameWithType);
+
+    watermarkForTable.compute(mergeLevel, (k, v) -> {
+      if (v == null) {
+        controllerMetrics.addCallbackGaugeIfNeeded(getMetricNameForTaskDelay(tableNameWithType, mergeLevel),
+            (() -> getMergeRollupTaskDelayInNumTimeBuckets(watermarkForTable.getOrDefault(k, -1L), bufferTimeMs,
+                bucketTimeMs)));
+      }
```

Comment:
Do we need add a `else` block here which remove the old CallbackGauge and add a new CallbackGauge when `v != null`?

Reply:
Our gauge will keep computing `getMergeRollupTaskDelayInNumTimeBuckets(watermarkForTable.getOrDefault(k, -1L), bufferTimeMs, bucketTimeMs))`. So, the gauge value will be changed if we just update the watermark data in `_mergeRollupWatermarks`. We don't need extra logic for remove & add the gauge.